### PR TITLE
When clauses + Specification for state change is implemented

### PR DESCRIPTION
**Pull request by d1jang**
@Ptival @ztatlock I've added a very immature feature that supports the when clause and specification for state change. 

To see what features are added, see ./test/when-00/kernel.krn, and try to compile it down to a Coq file.
See how ValidExchange/AddedValidExchange/KTrace are changed, and you'll get better understanding. See also exchange_ too. Traces are now tracking state change. For example, if a handler changes the value of counter by incremending by 1, and send it, it is reflected in the spec too.

I think my change is quite unstable, but I think we need these features ASAP, that's why I'm issuing a pull request now, so that we can stablize it.

### CODE REVIEW

**Comment by ztatlock** [comment]
Why do you introduce "state" to include in KTrace?

Can we have KTrace : kstate -> Trace -> Prop ?

**Comment by d1jang** [comment]
Sure, we can. I didn't like to pass around kstate which has many other things inside it including [Trace]. If we want to pass around kstate, isn't it better to have the type of KTrace to be just kstate->kstate->Prop? 

Moreover, I tried to use KTrace : kstate -> Trace -> Prop first, but the not discharged proof goals were very complicated and our sep foldr simplr tactics was not working with it that well, so I just kept using state outside kstate.

**Comment by ztatlock** [comment]
Just so that we're all on the same page:

We're going to use @d1jang's idea, but change KTrace to KInvariant and give it type:

``` ocaml
  KInvariant : kstate -> Prop
```

and furthermore change ValidExchange to have type:

``` ocaml
  ValidExchange : kstate -> kstate -> Prop
```

(where the first kstate is implemented as a parameter and not a type index).

Do you think we can do this tomorrow @d1jang? Let us know how we can support you!

This is the most important thing going on right now because it is a CORE part of the compiler and EVERYTHING else will depend on it!

**Comment by d1jang** [comment]
Sounds good as we discussed this afternoon!

I'll try to make the change tomorrow, first to an example code and next to the compiler itself.

I think @ztatlock and @Ptival can help when I get stuck with coming up with new simplr tactics that can discharge all the subgoals even with the changes I'll make. Let's see.

**Comment by ztatlock** [comment]
Awesome @d1jang! Happy to help with any part of this or chat about details of how the existing compiler works or etc. :)

**Comment by d1jang** [comment]
I've made a simple change in Kernel.v compiled from when-00 so that the types for the kernel spec are the same as you suggested. 

Specs are beautiful! 
But "sep foldr simplr" tactics now leaves some undischarged goals, and I don't know how to discharge them automatically.

I'm trying to figure it out, but I'm sure that we can make it work if we work together on this.
I'll send you the problematic code through email.

Let's make it work!!!

**Comment by ztatlock** [comment]
AWESOME Don! I will be offline until this evening due to this thing I committed to months ago, and the fact that Pat is in town. This is super important though and I will get to it ASAP! (Sorry that this weekend has been so busy :\ ...)

**Comment by d1jang** [comment]
The changes I made didn't take much time!! The real problem starts from now on.. I'll try to figure that out myself today. Let's cross our fingers :-)

**Comment by d1jang** [comment]
I tried to develop tactics which discharge all the subgoals for exchange_Main in Kernel.v where the KInvariant changes were made, but haven't finished yet.

The problem here is that Coq seems working poorly when the subterm of the goal is a record type and that subterm has to be unified with some other terms. Search for (\* working *) in the attached Kernel.v file.

If this problem is solved, it seems like we can _at least_ discharge all the subgoals for when-00.

I strongly believe that you Coq experts can show me a great way to solve this problem.

BTW, what's the inhabits type doing here? I guess that it's used here to consider a value in Set as in Prop so that we can extract the value out at runtime. But if that's inhabits's goal, why don't we just declare that value in Prop so that we don't have to worry about it? Is it because we're sometimes using a value in the same type either for proof and computation? I'm not so sure...

I'll email you the Kernel.v file I've been working on!

**Comment by Ptival** [comment]
Thanks for the new file, will take a look at it.

The goal:
traced (RecvMsg c ReqCounterValue ++ x) ==> traced x0
has left me perplex this day...

I'm not sure what causes that goal to appear, and how it did not appear in
other exchanges.

I'll let @ztatlock answer for inhabits, I'm not certain myself.
- Valentin

**Comment by Ptival** [comment]
BTW, @d1jang , you got rid of that goal I mentioned. Where was it coming from?

**Comment by d1jang** [comment]
The sep tactics couldn't discharge "traced (RecvMsg c ReqCounterValue ++ x) ==> traced x0", 
since it failed to derive the goal from the context that 

H0: tr = [x]
H1: (tr ~~~ RecvMsg c ReqCounterValue ++ tr) = [x0]%inhabited

The last two patterns in simplr are for the goal above. Doing unfold kst' in \* right before sep was very helpful for sep to figure out the relationships between many terms in the context, and it even could discharge the goal above. So we don't need the last two patterns in simplr anymore with the unfold kst' magic.

I described it as magic since I don't fully understand how sep is working :-(

**Comment by Ptival** [comment]
Here is an updated version of unfoldr that does its job better:

``` Coq
Ltac unfoldr :=
  repeat first [
    progress (unfold kstate_inv)
    | match goal with [ kst : kstate |- _ ] => unfold kst in * end
  ].
```

**Comment by Ptival** [comment]
@d1jang wants this merged, one of @ztatlock and @Ptival will have to read this.

File: compiler/src/genCoq.ml [hunk 1]
```diff
@@ -45,6 +50,20 @@ let rec coq_of_expr = function
     mkstr "(num_of_nat ((nat_of_num (%s)) + (nat_of_num (%s))))"
       (coq_of_expr a) (coq_of_expr b)
```
Comment (ztatlock) [inline compiler/src/genCoq.ml:55]: Please write some comments explaining eftable: what it is and how it is used.

File: compiler/src/kernel.ml [hunk 2]
```diff
@@ -14,6 +14,9 @@ type expr =
   | StrLit of string
   | Plus of expr * expr
 
+type lexpr =
```
Comment (ztatlock) [inline compiler/src/kernel.ml]: Let's rename lexpr to "when_cond".

File: compiler/src/kernel.ml [hunk 3]
```diff
@@ -41,14 +44,24 @@ type prog =
   | Nop
   | Seq of cmd * prog
 
+type taggedprog =
```
Comment (ztatlock) [inline compiler/src/kernel.ml]: Let's rename taggedprog to "cond_prog"

File: compiler/src/lex.mll [hunk 4]
```diff
@@ -20,12 +20,14 @@ rule token = parse
   | "Messages" { MESSAGES }
   | "Init" { INIT }
   | "Exchange" { EXCHANGE }
+  | "When" { WHEN }
```
Comment (ztatlock) [inline compiler/src/lex.mll]: "When" ==> "when"

**Comment by ztatlock** [comment]
Looks good. A few last TODOs:

(1) merge in master

(2) rename re: comments I made on code

(3) get the when test kernel to work (when-00)

(4) get spawn test to go through

**Comment by d1jang** [comment]
effecttable maintains an expression that represents the most recent value of a state variable. For exmaple, if an handler executes the following sequence of assignments :

1) a = b
2) a = a + 1
3) a = b + c

The effecttable is updated as the compiler encounters each of the assignments as follows

1) (a, b) :: nil
2) (a, b + 1) :: (a, b) :: nil
3) (a, b + 1 + c) :: (a, b + 1) :: (a,b) :: nil

This table is used to know how to represent the post condition of a state variable after 
executing a handler. So above, with state variables a b c as pre-condition, after executing
the handler, we get  (b + 1 + c) as the new value for a.

**Comment by Ptival** [comment]
This looks good. Now @d1jang, you will need to merge master into this branch:

``` git
git checkout master
git pull
git checkout whenclause
git merge master
```

At that point, you will probably have to fix some conflicts.

``` git
git status
```

should tell you which files are problematic. Personally, I use vimdiff, the default editor when doing:

```
git mergetool
```

It opens a vim with 4 panes, the bottom one being the "merged with conflicts" file that you need to clean up. At the top left and top right are the versions of this file from the two conflicting branches. You'll see things like this:

```
<<<<
blablabla
====
bliblibli
>>>>
```

For each such thing, you need to delete the lines you don't want (here, you could keep the "blablabla" or the "bliblibli" line, or just replace everything by "foofoofoo" if neither is okay.

Once you are done, you just need to commit this as you would commit a normal commit.

At any point, if you feel overwhelmed, stop and ask questions on what to do.

**Comment by d1jang** [comment]
Thanks for the information. I'm doing that now.. It seems like getCoq.ml is quite hard to merge now. 
I'm using the ediff mode for emacs. What tool are you doing for that?

**Comment by d1jang** [comment]
@Ptival After merging, what test do I have to run to see whether the merged one doesn't break your change?

**Comment by d1jang** [comment]
Now the master branch has been merged into then whenclause branch, and all tests can be just "COMPILED", but still proof cannot be done by the merged simplr tactics. @Ptival @ztatlock Can you guys take a look at why there's the rewriting problem I mentioned in the email, and emp ==> [XXX] cases are still remaining undischarged? 

I think this branch is almost there to be pulled to the master branch!! Hang in there, my awesome comrades!!!

**Comment by d1jang** [comment]
I think this branch is ready to be pulled to master. Now the merged compiler can compile all the test cases including u-quark and proofs are done automatically.

@ztatlock @Ptival Can you take a look at the code??

**Comment by Ptival** [comment]
@d1jang I am looking at it and want to commit a few changes before
proofreading it again for merge.

**Comment by ztatlock** [comment]
Awesome! I'lllook at this as soon as possible!

**Comment by Ptival** [comment]
@d1jang As a general remark, could you please try to set up your editor so that at least it expands tabs into sequences of spaces? Your code is unreadable with my tab width, and tabs are just annoying.

Even better would be for you to setup something to remove trailing whitespaces when you save a file. These are less annoying but still.

Your tactics are still very hard to read. You should probably document them next time they get this big, because what they are trying to achieve is fairly obscure until you try them manually on goals...

**Comment by d1jang** [comment]
@Ptival Next time you get annoyed by that kind of matter, please let me know first so that you don't have to read ill-formatted code. You can reformat the code if you want. If you didn't, push your changes to the remote branch so that I can work on that.

@ztatlock Can you send me your emacs configuration file so that I can format the code in the same way you're doing? I think that would be the best way to avoid these annoyances.

About tactics, I'll go and add some comments, but still WE have to work TOGETHER to make it better. I'm not still familiar with Ltac and how the sep tactic works, so maybe I've done something unnecessary there.

**Comment by d1jang** [comment]
@Ptival One more thing. How trailing whitespeaces are annoying you? Is it because they are not necessary or do them cause annoying visual effects in your editor?

**Comment by Ptival** [comment]
@d1jang I've done some changes on my side to remove and reformat the code as needed, will push that at some point today.

For the trailing whitespaces, it's more of a minor remark because they are useless, they don't do anything bad, unlike tabs.

We can make sure everything is okay and perform the merge tomorrow morning.

File: compiler/src/genCoq.ml [hunk 5]
```diff
@@ -48,6 +53,23 @@ let rec coq_of_expr = function
...
 
+(* eftable : a variname name -> an expressions containing the current
```
Comment (Ptival) [inline compiler/src/genCoq.ml]: What's a variname name!?

**Comment by Ptival** [comment]
I tried to simplify things a bit. I understand now that it was not too easy to get things right, given how goals sometimes look the same but need a fair amount of preparation to flow through sep. 3b4fdac passes all tests.

**Comment by Ptival** [comment]
Ok, @d1jang you can merge.

**Pull request merged**